fmp_server.py

```diff
@@ -898,6 +906,19 @@ def sendAnsCircleMatchEnd(self, seq):
         """
         self.send_packet(PatID4.AnsCircleMatchEnd, b"", seq)
 
+    def handle(self):
+        PatRequestHandler.handle(self)  # "super()" does not work with old-style classes
+        if self.session.graceful_shutdown is False:
+            if self.session.layer == 2:
+                new_host = self.session.try_transfer_city_leadership()
+                if new_host:
+                    self.sendNtcLayerHost(new_host, 0)
+            if self.session.layer > 0:
+                ntc_data = pati.lp2_string(self.session.capcom_id)
+                self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
+                                            ntc_data, 0)
+            self.session.shutdown(1)
+        self.session.graceful_shutdown = False
```

Comment:
I don't think it's useful as it will be set to `False` again on the next connection.

Reply:
I thought so too, and it took me a little bit to figure this out, but it breaks if this line is not there. I can only assume that the "master" PatHandler class that this project inherits from has some underlying logic that inherits the old `Session`.

Reply:
Ah, that's right. Rather than implementing it to the `ReqConnection` packet, you should instead implement it in the `session.get` method instead. Though, as I've mentioned, I'm pretty sure it can be dealt with by cleaning it up in the `except` block of the `handle` method.

Reply:
In my local wip branch I changed this part: https://github.com/sepalani/MH3SP/blob/703cc1c086a6e181ed474776fb00c6430eaa4cc0/mh/session.py#L69

I added this guard:
```python
session = DB.get_session(self)
if session != self:
    assert session.connection is None, "Session is already in use"
    session.connection = self.connection
    self.connection = None
return session
```